pkg/authentication/authenticate.go

```diff
@@ -0,0 +1,209 @@
+	http.StripPrefix("/auth", a.router).ServeHTTP(w, r)
+}
+
+// Authenticate starts a new authentication (by redirecting the user to the Login UI)
+// The initially requested URI (in the application) is passed as encrypted state.
+func (a *Authenticator[T]) Authenticate(w http.ResponseWriter, r *http.Request, requestedURI string) {
+	s := &State{RequestedURI: requestedURI}
+	stateParam, err := s.Encrypt(a.encryptionKey)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	a.authN.Authenticate(w, r, stateParam)
+}
+
+// Callback handles the redirect back from the Login UI. On successful authentication a new session
+// will be created and its id will be stored in a cookie.
+// The user will be redirected to the initially requested UI (passed as encrypted state)
+func (a *Authenticator[T]) Callback(w http.ResponseWriter, req *http.Request) {
+	ctx, stateParam := a.authN.Callback(w, req)
+	if !ctx.IsAuthenticated() {
+		a.logger.Error("unauthenticated after callback")
+		http.Error(w, "not authenticated", http.StatusForbidden)
+		return
+	}
+	state, err := DecryptState(stateParam, a.encryptionKey)
+	if err != nil {
+		a.logger.Error("unable to decrypt state", "state", stateParam)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	id := uuid.NewString()
+	err = a.setSessionCookie(w, id)
+	if err != nil {
+		a.logger.Error("unable to save session cookie", "error", err, "id", id)
+		http.Error(w, "session could not be stored", http.StatusInternalServerError)
+		return
+	}
+	err = a.sessions.Set(id, ctx)
+	if err != nil {
+		a.logger.Error("unable to save session", "error", err, "id", id)
+		http.Error(w, "session could not be stored", http.StatusInternalServerError)
+		return
+	}
+
+	http.Redirect(w, req, state.RequestedURI, http.StatusFound)
+}
+
+// Logout will terminate the exising session.
+func (a *Authenticator[T]) Logout(w http.ResponseWriter, req *http.Request) {
+	ctx, err := a.IsAuthenticated(req)
+	if err != nil {
+		http.Redirect(w, req, "/", http.StatusFound)
+		return
+	}
+	s := &State{RequestedURI: ""}
+	stateParam, err := s.Encrypt(a.encryptionKey)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	a.deleteSessionCookie(w)
+
+	proto := "http"
+	if req.TLS != nil {
+		proto = "https"
+	}
+	postLogout := fmt.Sprintf("%s://%s/", proto, req.Host)
+	a.authN.Logout(w, req, ctx, stateParam, postLogout)
+}
+
+// IsAuthenticated checks whether there is an existing session of not.
+// In case there is one, it will be returned.
+func (a *Authenticator[T]) IsAuthenticated(req *http.Request) (T, error) {
+	var t T
+	cookie, err := req.Cookie(a.sessionCookieName)
+	if err != nil {
+		return t, ErrNoCookie
+	}
+	sessionID, err := crypto.DecryptAES(cookie.Value, a.encryptionKey)
+	if err != nil {
+		a.logger.Log(req.Context(), slog.LevelWarn, "unable to decrypt session cookie", "cookie value", cookie.Value)
+		return t, ErrNoSession
```

Comment:
Additional error like "ErrCryptFailed"?

Reply:
the intention was to return as minimal as needed...
no cookie so there can't be a session or the cookie provided did not match to a session